trl/trainer/sft_trainer.py

```diff
@@ -358,6 +364,7 @@ def _prepare_dataset(
         packing: bool,
         formatting_func: Optional[Callable[[dict], str]],
         dataset_name: str,
+        pack_smart: bool = False,
```

Comment:
maybe this is better:

```suggestion
        packing: Union[bool, str],
        formatting_func: Optional[Callable[[dict], str]],
        dataset_name: str,
```

and use `if packing == "knapsack":`

Reply:
fixed!